### PR TITLE
feat: Add `get_chain_root_span` utility for langchain instrumentation

### LINKED ISSUE #1052: [feature request] Add an easy way to get the root span from a child span
**Is your feature request related to a problem? Please describe.**
I want to implement a feedback sending mechanism from the front-end of a chat bot to a **python** backend.
For LangChain, it is possible to add a link in the chain that fetches the current span_id using `from openinference.instrumentation.langchain import get_current_span`, but this only gives the span_id of that link.
It is much more natural to get the root span id, which is the correct one to annotate with the user-sent feedback, but it's currently not easy to accomplish programmatically. See discussion here: https://github.com/Arize-ai/phoenix/discussions/4800

**Describe the solution you'd like**
An easy to use mechanism to get the root span_id of the currently running trace, ideally in a way that is agnostic to the currently instrumented LLM framework solution, but also a specific one for each framework would be nice.

**Describe alternatives you've considered**
I've considered creating the root span manually and then calling a chain within that with the automatically instrumented langchain, but that is very inelegant.

### PR DESCRIPTION
**Pull request by anticorrelator**
partially resolves #4158
resolves #1052 

Adds a utility to get the most recent chain span that can be found traversing up the trace tree. This will help grab the most appropriate span for sending user feedback.

### CODE REVIEW

File: python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/_tracer.py
Comment (RogerHYang) [inline python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/_tracer.py:150]: 1. What are the benefits of modifying the span, versus keeping this information external to the span, e.g. in a tuple or a custom object?
2. What if in the future users want a different `run_type`? How adaptable is this approach: do we need to reserve additional attribute keywords for this purpose?

File: python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/_tracer.py
Comment (RogerHYang) [inline python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/_tracer.py:37]: What are the benefits of depending on the SDK version of the `Span` instead of the abstract version of from the Trace API? Which specific implementation detail of the SDK are we relying on?

File: python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/_tracer.py
Comment (RogerHYang) [inline python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/_tracer.py:109]: Although it probably makes no functional difference, but regarding code organization, what are the benefits of making this a module variable (instead of an instance variable)?

File: python/instrumentation/openinference-instrumentation-langchain/tests/test_instrumentor.py
Comment (anticorrelator) [reply python/instrumentation/openinference-instrumentation-langchain/tests/test_instrumentor.py:124]: Yeah, I briefly had one of the tests use a `RunnableSequence` instead but it was really hard for me to figure out how to satisfy mypy simultaneously for both the 3.9 and 3.12 run. I can keep working at it if you think it's worth it

File: python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/_tracer.py
Comment (anticorrelator) [reply python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/_tracer.py:37]: The abstract `Span` object doesn't have the `attributes` or `parent` attributes defined on the object, so our ancestor-walking method will fail type checks without using the fully specified sdk Span

File: python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/_tracer.py
Comment (anticorrelator) [reply python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/_tracer.py:109]: it's primarily to make it easily definable in the `__init__` module without needing to retrieve the active instrumentor

File: python/instrumentation/openinference-instrumentation-langchain/tests/test_instrumentor.py
Comment (RogerHYang) [reply python/instrumentation/openinference-instrumentation-langchain/tests/test_instrumentor.py:124]: If it's just an issue with typing, we could do `type: ignore` since static typing is not a primary concern for testing purposes.

Comment (anticorrelator) [reply python/instrumentation/openinference-instrumentation-langchain/tests/test_instrumentor.py:124]: the type: ignore works (I think) for 3.12 checks but for 3.9 checks I get an "unused type:ignore" error, happy to try again but I felt a little bit like it was unnecessary

File: python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/_tracer.py
Comment (RogerHYang) [reply python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/_tracer.py:109]: The otel instrumentor is a singleton, so calling `LangChainInstrument()` will work, as shown [here](https://github.com/Arize-ai/openinference/blob/ff43e9ddc0a5f683f80d09139247ad194d6c29af/python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/__init__.py#L106).

File: python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/_tracer.py
Comment (RogerHYang) [reply python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/_tracer.py:37]: Since the parent relations can be obtain from the (abstract) `get_span_context` method, the only reason to depend on the SDK is because we need to write to and read back values from the span attributes, is that correct?

Comment (RogerHYang) [reply python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/_tracer.py:37]: Since both of these use cases can be mitigated by having our own bookkeeping, ideally we could depend only on the API and no the SDK, so that we can keep the instrumentor agnostic regarding the `tracer_provider` that's given.

Comment (RogerHYang) [reply python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/_tracer.py:37]: On the other hand, if you don't think it's worthwhile to abide by just the API, then we should add a validation at instantiation time to confirm the SDK implementation, so there's no ambiguity at the user interface level for the instrumentor.

File: python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/_tracer.py
Comment (anticorrelator) [reply python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/_tracer.py]: I moved all the bookkeeping onto the tracer

File: python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/_tracer.py
Comment (anticorrelator) [reply python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/_tracer.py]: I moved all the bookkeeping to the tracer, so I reverted this to use the abstract API

File: python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/__init__.py
Comment (RogerHYang) [inline python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/__init__.py:78]: redundant logic
```suggestion
```

File: python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/__init__.py
Comment (RogerHYang) [inline python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/__init__.py:67]: I think by making the following signature change, we could make this more general and at the same time simplify our internal logic, given the following compositional equivalence. This simplifies our backend logic because we only need a `Dict[UUID, Optional[UUID]]` to track whose parent is whom, and we can just call on the preexisting `self.run_map`.
```python
get_root_chain_span(run_id) is get_span(get_root_chain_run_id(run_id))
```
We just have to also update `get_span` to take `None` as argument, which is a trivial change.
```suggestion
    def get_root_chain_run_id(self, run_id: UUID) -> Optional[UUID]:
```

File: python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/__init__.py
Comment (RogerHYang) [inline python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/__init__.py:76]: Strictly speaking, we should not be using private methods in this scope.

File: python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/__init__.py
Comment (RogerHYang) [inline python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/__init__.py:136]: If you have time, would you mind DRY-ing this up, since it's the same as line 107-119. Thx!

File: python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/__init__.py
Comment (RogerHYang) [reply python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/__init__.py:67]: FYI, [`self.run_map`](https://github.com/langchain-ai/langchain/blob/2a1029c53c9c14b177eb61743dc0501f97dd0738/libs/core/langchain_core/tracers/core.py#L117) seems to be all you need.

File: python/instrumentation/openinference-instrumentation-langchain/tests/test_instrumentor.py
Comment (anticorrelator) [reply python/instrumentation/openinference-instrumentation-langchain/tests/test_instrumentor.py:157]: The RunnableLambdas create their own chain, so they are their own root

File: python/instrumentation/openinference-instrumentation-langchain/tests/test_instrumentor.py
Comment (anticorrelator) [reply python/instrumentation/openinference-instrumentation-langchain/tests/test_instrumentor.py:147]: the callables are entirely blocking anyway, I don't think gather does anything different

File: python/instrumentation/openinference-instrumentation-langchain/tests/test_instrumentor.py
Comment (RogerHYang) [reply python/instrumentation/openinference-instrumentation-langchain/tests/test_instrumentor.py:157]: > The RunnableLambdas create their own chain, so they are their own root

If that's the case then this is not really testing the tree climbing procedure. Is that correct?

Comment (anticorrelator) [reply python/instrumentation/openinference-instrumentation-langchain/tests/test_instrumentor.py:157]: It's not, though if like you suggested we yield the the ancestor chains we can test how many there are. I didn't end up doing that since it feels like a confusing interface?

Comment (RogerHYang) [reply python/instrumentation/openinference-instrumentation-langchain/tests/test_instrumentor.py:157]: This is not related to yielding the ancestor chain. All i'm pointing out is that we are _not_ testing what we have implemented.

File: python/instrumentation/openinference-instrumentation-langchain/tests/test_instrumentor.py
Comment (RogerHYang) [reply python/instrumentation/openinference-instrumentation-langchain/tests/test_instrumentor.py:147]: Do you want to add a sleep like we have done in an [existing test](https://github.com/Arize-ai/openinference/blob/c3ce79402a05e99640b905274a16a8c0af8c04a2/python/instrumentation/openinference-instrumentation-langchain/tests/test_instrumentor.py#L76)? I'm just pointing out that f we aren't testing it concurrently then there is no need to have `n` of them: a single one should suffice for a test that is serial.

File: python/instrumentation/openinference-instrumentation-langchain/tests/test_instrumentor.py
Comment (anticorrelator) [reply python/instrumentation/openinference-instrumentation-langchain/tests/test_instrumentor.py:157]: I know that, I'm suggesting yielding the ancestor chain can be an easy way to do that

Comment (RogerHYang) [reply python/instrumentation/openinference-instrumentation-langchain/tests/test_instrumentor.py:157]: > I know that

So are you suggesting that the logic you have implemented so far doesn't need to be tested?

File: python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/__init__.py
Comment (RogerHYang) [reply python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/__init__.py:136]: I'm just pointing out that this is a simple copy-paste. I'm not suggesting adding abstractions or anything like that. But it's up to you.

File: python/instrumentation/openinference-instrumentation-langchain/tests/test_instrumentor.py
Comment (RogerHYang) [reply python/instrumentation/openinference-instrumentation-langchain/tests/test_instrumentor.py:157]: I’m not saying it has to be tested, but I’ve noticed the logic isn’t being tested. Since we seem to agree on that observation, I’d like to understand why leaving it untested is considered the better option.

Comment (anticorrelator) [reply python/instrumentation/openinference-instrumentation-langchain/tests/test_instrumentor.py:157]: I am specifically offering an alternative strategy to testing whether or not it's the right ancestor directly: if we change this implementation to yield ancestors we can do (I think) a good enough test by counting how many ancestors are yielded each time instead of directly checking the ancestor directly.

Regarding needing to test it: I can go either way on it, I'm not positive this test is needed though I agree it would be nice. I'm using tests as a guide here to scaffold the implementation, not necessarily be exhaustive.

Comment (RogerHYang) [reply python/instrumentation/openinference-instrumentation-langchain/tests/test_instrumentor.py:157]: > I'm not positive this test is needed though I agree it would be nice.  I'm using tests as a guide here to scaffold the implementation, not necessarily be exhaustive.

While I understand the test may not seem essential right now, it’s important to consider that this repo is updated only occasionally. Given the infrequent changes and the high likelihood that any future updates may be handled by someone new, having tests that are both thorough and precise can be incredibly beneficial. These tests can serve as reliable documentation and help prevent potential issues when the codebase is revisited later.

> I am specifically offering an alternative strategy to testing 

As an aside, I previously mentioned that the self.run_map from the base instrumentor already stores the family tree of UUIDs. You can leverage this to simplify both the implementation and the testing, as it’s the ultimate source of truth for the family tree.

File: python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/__init__.py
Comment (RogerHYang) [reply python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/__init__.py:136]: > my preference for first implementations

I’m suggesting that if you don’t anticipate making further updates to this repo in the near future, it’s generally a good practice to replace the copy-pasted sections with a helper function. However, if you think it’s better to leave that for the next maintainer, that’s also fine.

File: python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/__init__.py
Comment (RogerHYang) [inline python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/__init__.py:67]: Unless you disagree, “root” typically refers to just a single node, so naming the function that way would be misleading if it intends to return a list. Based on my reading, this function returns all ancestors that are chains, and skips those that are not, so if the final root of the tree is not a `chain`, it is actually not returned.

File: python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/__init__.py
Comment (RogerHYang) [inline python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/__init__.py:78]: nit. might as well just return an empty list, since it makes no practical difference but simplifies the types.
```suggestion
        return root_chain_spans
```

File: python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/__init__.py
Comment (RogerHYang) [inline python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/__init__.py:74]: As mentioned earlier, in this function you could use the existing `tracer.run_map` from the base `Tracer` to simplify your implementation, rather than accessing private attributes in this scope. Since `tracer.run_map` already provides all the necessary functionality, there’s no need to re-implement everything from scratch: the parts highlighted by arrows in the screenshot below would give you the [`parent_run_id`](https://github.com/langchain-ai/langsmith-sdk/blob/5e21109e1cd70ee19db29941c128223ef360a40b/python/langsmith/schemas.py#L254) and [`run_type`](https://github.com/langchain-ai/langsmith-sdk/blob/5e21109e1cd70ee19db29941c128223ef360a40b/python/langsmith/schemas.py#L225) that you're using for your tree traversal algorithm.

Screenshot 2024-10-16 at 8 11 51 AM

**Review by RogerHYang — APPROVED** [review]
Approved to unblock. Concerns are noted in comments

File: python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/__init__.py
Comment (mikeldking) [inline python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/__init__.py:67]:
```suggestion
    def get_ancestor_spans(self, run_id: UUID) -> Optional[List[Span]]:
```
just to be explicit about the return type

File: python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/__init__.py
Comment (mikeldking) [inline python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/__init__.py:85]: I think for the simplicity of the typing (and maybe just my FP brain) - I think an empty array might be preferable to represent "none"?

**Review by mikeldking — APPROVED** [review]

**Pull request merged**
